editor/xml_feature.cpp

```diff
@@ -0,0 +1,401 @@
+#include "editor/xml_feature.hpp"
+
+#include "base/assert.hpp"
+#include "base/string_utils.hpp"
+#include "base/timer.hpp"
+
+#include "geometry/mercator.hpp"
+
+#include "std/set.hpp"
+#include "std/unordered_set.hpp"
+
+#include <boost/functional/hash.hpp>
+
+#include "3party/pugixml/src/pugixml.hpp"
+
+namespace
+{
+auto constexpr kLatLonTolerance = 7;
+
+pugi::xml_node FindTag(pugi::xml_document const & document, string const & key)
+{
+  return document.select_node(("//tag[@k='" + key + "']").data()).node();
+}
+
+m2::PointD PointFromLatLon(pugi::xml_node const node)
+{
+  double lat, lon;
+  if (!strings::to_double(node.attribute("lat").value(), lat))
+    MYTHROW(editor::XMLFeatureNoLatLonError,
+            ("Can't parse lat attribute: " + string(node.attribute("lat").value())));
+ if (!strings::to_double(node.attribute("lon").value(), lon))
+   MYTHROW(editor::XMLFeatureNoLatLonError,
+           ("Can't parse lon attribute: " + string(node.attribute("lon").value())));
+  return MercatorBounds::FromLatLon(lat, lon);
+}
+
+void ValidateNode(pugi::xml_node const & node)
+{
+  if (!node)
+    MYTHROW(editor::XMLFeatureNoNodeError, ("Document has no node"));
+
+  try
+  {
+    PointFromLatLon(node);
+  }
+  catch(editor::XMLFeatureError const &)
+  {
+    throw;
```

Comment:
What's the point in try/catch at all, if you explicitly forwards XML exceptions here.